src/Plutarch/Extra/Ord.hs

```diff
@@ -0,0 +1,641 @@
+                                    PGT -> pcons # ysHead #$ passertOrderBy # cmp # xs
+                                    _ -> pcons # xsHead #$ self # cmp # xsTail # ys
+                                -- Need to look further to ensure we're not mis-ordered
+                                PJust ys'' -> pmatch xs'' $ \(PPair xsPeek _) ->
+                                    -- Verify by look-ahead that xsHead and xsPeek are ordered
+                                    pmatch (pcompareBy # cmp # xsHead # xsPeek) $ \case
+                                        -- We are out of order, vomit.
+                                        PGT -> ptraceError "ptryMergeBy: argument list-like out of order"
+                                        _ -> pmatch ys'' $ \(PPair ysPeek _) ->
+                                            -- Verify by look-ahead that ysHead and ysPeek are ordered
+                                            pmatch (pcompareBy # cmp # ysHead # ysPeek) $ \case
+                                                -- We are out of order, vomit.
+                                                PGT -> ptraceError "ptryMergeBy: argument list-like out of order"
+                                                -- Actually perform the merge.
+                                                _ -> pmatch (pcompareBy # cmp # xsHead # ysHead) $ \case
+                                                    PGT -> pcons # ysHead #$ self # cmp # xs # ysTail
+                                                    _ -> pcons # xsHead #$ self # cmp # xsTail # ys
+
+{- | Verifies if every element of the input list-like is unique according to its
+ 'PEq' instance. Relies on a 'POrd' instance for speed, but only works if the
+ input is sorted according to that instance.
+
+ Returns @'PJust' 'PTrue'@ if every element is unique as specified above,
+ @'PJust' 'PFalse'@ if a duplicate exists, and 'PNothing' if the input isn't
+ ordered.
+
+ @since 3.5.0
+-}
+pAllUnique ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (PElemConstraint ell a, PListLike ell, POrd a) =>
+    Term s (ell a :--> PMaybe PBool)
+pAllUnique = phoistAcyclic $ plam $ \xs -> pAllUniqueBy # pfromOrd @a # xs
+
+{- | As 'pAllUnique', but relies on a custom 'PComparator' for both equality and
+ ordering (instead of a 'PEq' and 'POrd' instance respectively).
+
+ @since 3.5.0
+-}
+pAllUniqueBy ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (PElemConstraint ell a, PListLike ell) =>
+    Term s (PComparator a :--> ell a :--> PMaybe PBool)
+pAllUniqueBy = phoistAcyclic $
+    plam $ \cmp ->
+        precList (go cmp) (const (pcon . PJust . pconstant $ True))
+  where
+    go ::
+        forall (s' :: S).
+        Term s' (PComparator a) ->
+        Term s' (ell a :--> PMaybe PBool) ->
+        Term s' a ->
+        Term s' (ell a) ->
+        Term s' (PMaybe PBool)
+    go cmp self x xs = pmatch (PList.puncons # xs) $ \case
```

Comment:
This looks very inefficient. Not only does it peek ahead instead of remembering using an extra parameter of the recursive call, it also creates a `PMaybe` value for each element and matches on that.